src/System.Windows.Extensions/src/System/Media/SoundPlayer.cs

```diff
@@ -0,0 +1,738 @@
+
+                // Play the path - don't use uri.AbsolutePath because that gives problems
+                // when there are whitespaces in file names
+                IsLoadCompleted = true;
+
+                ValidateSoundFile(localPath);
+                Interop.WinMM.PlaySound(localPath, IntPtr.Zero, Interop.WinMM.SND_NODEFAULT | flags);
+            }
+            else
+            {
+                LoadSync();
+                ValidateSoundData(_streamData);
+                Interop.WinMM.PlaySound(_streamData, IntPtr.Zero, Interop.WinMM.SND_MEMORY | Interop.WinMM.SND_NODEFAULT | flags);
+            }
+        }
+
+        private void LoadSync()
+        {
+            Debug.Assert((_uri == null || !_uri.IsFile), "we only load streams");
+
+            // first make sure that any possible download ended
+            if (!_semaphore.WaitOne(LoadTimeout, false))
+            {
+                _copyThread?.Abort();
+                CleanupStreamData();
+                throw new TimeoutException(SR.SoundAPILoadTimedOut);
+            }
+
+            // if we have data, then we are done
+            if (_streamData != null)
+            {
+                return;
+            }
+
+            // setup the http stream
+            if (_uri != null && !_uri.IsFile && _stream == null)
+            {
+                WebRequest webRequest = WebRequest.Create(_uri);
+                webRequest.Timeout = LoadTimeout;
+
+                WebResponse webResponse;
+                webResponse = webRequest.GetResponse();
+
+                // now get the stream
+                _stream = webResponse.GetResponseStream();
+            }
+
+            if (_stream.CanSeek)
+            {
+                // if we can get data synchronously, then get it
+                LoadStream(true);
+            }
+            else
+            {
+                // the data can't be loaded synchronously
+                // load it async, then wait for it to finish
+                _doesLoadAppearSynchronous = true; // to avoid OnFailed call.
+                LoadStream(false);
+
+                if (!_semaphore.WaitOne(LoadTimeout, false))
+                {
+                    _copyThread?.Abort();
+                    CleanupStreamData();
+                    throw new TimeoutException(SR.SoundAPILoadTimedOut);
+                }
+
+                _doesLoadAppearSynchronous = false;
+
+                if (_lastLoadException != null)
+                {
+                    throw _lastLoadException;
+                }
+            }
+
+            // we don't need the worker copyThread anymore
+            _copyThread = null;
+        }
+
+        private void LoadStream(bool loadSync)
+        {
+            if (loadSync && _stream.CanSeek)
+            {
+                int streamLen = (int)_stream.Length;
+                _currentPos = 0;
+                _streamData = new byte[streamLen];
+                _stream.Read(_streamData, 0, streamLen);
```

Comment:
This is broken.  The code is assuming that Read will always read the requested number of bytes... that's not the case.